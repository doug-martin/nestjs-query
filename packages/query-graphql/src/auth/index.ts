export * from './authorizer';
export * from './default-crud.authorizer';
export * from './tokens';
