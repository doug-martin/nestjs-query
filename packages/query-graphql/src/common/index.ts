export { DTONamesOpts, getDTONames, DTONames, getDTOIdTypeOrDefault } from './dto.utils';
export * from './external.utils';
export * from './resolver.utils';
export * from './object.utils';
