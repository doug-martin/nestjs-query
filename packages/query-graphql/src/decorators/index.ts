export {
  FilterableField,
  FilterableFieldOptions,
  FilterableFieldDescriptor,
  getFilterableFields,
} from './filterable-field.decorator';
export { ResolverMethodOpts } from './resolver-method.decorator';
export {
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  Relation,
  FilterableRelation,
  RelationDecoratorOpts,
  RelationTypeFunc,
  getRelations,
} from './relation.decorator';
export * from './resolver-mutation.decorator';
export * from './resolver-query.decorator';
export * from './resolver-field.decorator';
export { Reference, ReferenceDecoratorOpts, ReferenceTypeFunc } from './reference.decorator';
export { ResolverSubscription, SubscriptionResolverMethodOpts } from './resolver-subscription.decorator';
export { InjectPubSub } from './inject-pub-sub.decorator';
export * from './skip-if.decorator';
export * from './aggregate-query-param.decorator';
export * from './hook.decorator';
export * from './decorator.utils';
export * from './hook-args.decorator';
export * from './authorizer.decorator';
export * from './inject-authorizer.decorator';
export * from './inject-custom-authorizer.decorator';
export * from './key-set.decorator';
export * from './authorize-filter.decorator';
export * from './query-options.decorator';
export * from './id-field.decorator';
