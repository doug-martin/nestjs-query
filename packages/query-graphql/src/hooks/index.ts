export * from './hooks';
export { createDefaultHook } from './default.hook';
export * from './tokens';
export * from './types';
