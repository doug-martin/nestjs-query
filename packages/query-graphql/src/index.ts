export * from './types';
export {
  FilterableField,
  FilterableFieldOptions,
  ResolverMethodOpts,
  Relation,
  FilterableRelation,
  CursorConnection,
  FilterableCursorConnection,
  OffsetConnection,
  FilterableOffsetConnection,
  UnPagedRelation,
  FilterableUnPagedRelation,
  RelationTypeFunc,
  RelationDecoratorOpts,
  Reference,
  ReferenceTypeFunc,
  ReferenceDecoratorOpts,
  InjectPubSub,
  BeforeCreateOne,
  BeforeCreateMany,
  BeforeUpdateOne,
  BeforeUpdateMany,
  BeforeDeleteOne,
  BeforeDeleteMany,
  BeforeQueryMany,
  BeforeFindOne,
  InjectAuthorizer,
  InjectCustomAuthorizer,
  Authorize,
  AuthorizerFilter,
  RelationAuthorizerFilter,
  HookArgs,
  MutationHookArgs,
  KeySet,
  QueryOptions,
  IDField,
  IDFieldOptions,
} from './decorators';
export * from './resolvers';
export * from './federation';
export { DTONamesOpts } from './common';
export { NestjsQueryGraphQLModule } from './module';
export { AutoResolverOpts } from './providers';
export { pubSubToken, GraphQLPubSub } from './subscription';
export { Authorizer, CustomAuthorizer, AuthorizerOptions, AuthorizationContext, OperationGroup } from './auth';
export {
  Hook,
  HookTypes,
  BeforeCreateOneHook,
  BeforeCreateManyHook,
  BeforeUpdateOneHook,
  BeforeUpdateManyHook,
  BeforeDeleteOneHook,
  BeforeDeleteManyHook,
  BeforeQueryManyHook,
  BeforeFindOneHook,
} from './hooks';
export { AuthorizerInterceptor, AuthorizerContext, HookInterceptor, HookContext } from './interceptors';
