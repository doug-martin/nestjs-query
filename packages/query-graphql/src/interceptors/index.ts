export * from './hook.interceptor';
export * from './authorizer.interceptor';
