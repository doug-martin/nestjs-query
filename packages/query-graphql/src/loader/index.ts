export * from './find-relations.loader';
export * from './query-relations.loader';
export * from './count-relations.loader';
export * from './aggregate-relations.loader';
export * from './dataloader.factory';
