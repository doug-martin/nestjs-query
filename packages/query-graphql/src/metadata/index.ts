import { GraphQLQueryMetadataStorage } from './metadata-storage';

export { FilterableFieldDescriptor } from './metadata-storage';

/** @internal */
let storage: GraphQLQueryMetadataStorage;
/** @internal */
export function getMetadataStorage(): GraphQLQueryMetadataStorage {
  if (!storage) {
    storage = new GraphQLQueryMetadataStorage();
  }
  return storage;
}
