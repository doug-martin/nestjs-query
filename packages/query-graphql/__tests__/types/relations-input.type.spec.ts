// eslint-disable-next-line max-classes-per-file
import { plainToClass } from 'class-transformer';
import { validateSync } from 'class-validator';
import { Args, Query, Resolver, Int, InputType, ObjectType } from '@nestjs/graphql';
import { FilterableField, IDField, RelationsInputType } from '../../src';
import { generateSchema } from '../__fixtures__';

describe('RelationsInputType', (): void => {
  @ObjectType()
  class ParentDTO {
    @FilterableField()
    field!: string;
  }

  @ObjectType()
  class ParentCustomIDDTO {
    @IDField(() => String)
    id!: string;
  }

  @ObjectType()
  class RelationDTO {
    @FilterableField()
    relationField!: string;
  }

  @ObjectType()
  class RelationCustomIDDTO {
    @IDField(() => String)
    relationId!: string;
  }

  @InputType()
  class RelationsInput extends RelationsInputType(ParentDTO, RelationDTO) {}

  it('should create an input type with an id and relationIds fields', async () => {
    @Resolver()
    class RelationsInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationsInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationsInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an input type with a custom parent id', async () => {
    @InputType()
    class RelationsCustomParentIdInput extends RelationsInputType(ParentCustomIDDTO, RelationDTO) {}

    @Resolver()
    class RelationsCustomIdInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationsCustomParentIdInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationsCustomIdInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an input type with a custom relation id', async () => {
    @InputType()
    class RelationsCustomRelationIdInput extends RelationsInputType(ParentDTO, RelationCustomIDDTO) {}

    @Resolver()
    class RelationsCustomIdInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationsCustomRelationIdInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationsCustomIdInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should create an input type with a custom parent and relation id', async () => {
    @InputType()
    class RelationsCustomParentRelationIdInput extends RelationsInputType(ParentCustomIDDTO, RelationCustomIDDTO) {}

    @Resolver()
    class RelationsCustomIdInputTypeSpec {
      @Query(() => Int)
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      test(@Args('input') input: RelationsCustomParentRelationIdInput): number {
        return 1;
      }
    }
    const schema = await generateSchema([RelationsCustomIdInputTypeSpec]);
    expect(schema).toMatchSnapshot();
  });

  it('should return the input when accessing the update field', () => {
    const input: RelationsInputType = { id: 1, relationIds: [2, 3, 4] };
    const it = plainToClass(RelationsInput, input);
    expect(it.id).toEqual(input.id);
    expect(it.relationIds).toEqual(input.relationIds);
  });

  describe('validation', () => {
    it('should validate the id is defined', () => {
      const input = { relationIds: [2, 3, 4] };
      const it = plainToClass(RelationsInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'id should not be empty',
          },
          property: 'id',
          target: input,
        },
      ]);
    });

    it('should validate the id is not empty', () => {
      const input = { id: '', relationIds: [2, 3, 4] };
      const it = plainToClass(RelationsInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'id should not be empty',
          },
          property: 'id',
          target: input,
          value: '',
        },
      ]);
    });

    it('should allow an empty relationIds array', () => {
      const input: RelationsInputType = { id: 1, relationIds: [] };
      const it = plainToClass(RelationsInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([]);
    });

    it('should validate that relationsIds is unique', () => {
      const input: RelationsInputType = { id: 1, relationIds: [1, 2, 1, 2] };
      const it = plainToClass(RelationsInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            arrayUnique: "All relationIds's elements must be unique",
          },
          property: 'relationIds',
          target: input,
          value: input.relationIds,
        },
      ]);
    });

    it('should validate that relationsIds does not contain an empty id', () => {
      const input: RelationsInputType = { id: 1, relationIds: [''] };
      const it = plainToClass(RelationsInput, input);
      const errors = validateSync(it);
      expect(errors).toEqual([
        {
          children: [],
          constraints: {
            isNotEmpty: 'each value in relationIds should not be empty',
          },
          property: 'relationIds',
          target: input,
          value: input.relationIds,
        },
      ]);
    });
  });
});
