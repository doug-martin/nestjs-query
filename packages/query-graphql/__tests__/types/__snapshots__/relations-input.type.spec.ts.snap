// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RelationsInputType should create an input type with a custom parent and relation id 1`] = `
type Query {
  test(input: RelationsCustomParentRelationIdInput!): Int!
}

input RelationsCustomParentRelationIdInput {
  """The id of the record."""
  id: String!

  """The ids of the relations."""
  relationIds: [String!]!
}

`;

exports[`RelationsInputType should create an input type with a custom parent id 1`] = `
type Query {
  test(input: RelationsCustomParentIdInput!): Int!
}

input RelationsCustomParentIdInput {
  """The id of the record."""
  id: String!

  """The ids of the relations."""
  relationIds: [ID!]!
}

`;

exports[`RelationsInputType should create an input type with a custom relation id 1`] = `
type Query {
  test(input: RelationsCustomRelationIdInput!): Int!
}

input RelationsCustomRelationIdInput {
  """The id of the record."""
  id: ID!

  """The ids of the relations."""
  relationIds: [String!]!
}

`;

exports[`RelationsInputType should create an input type with an id and relationIds fields 1`] = `
type Query {
  test(input: RelationsInput!): Int!
}

input RelationsInput {
  """The id of the record."""
  id: ID!

  """The ids of the relations."""
  relationIds: [ID!]!
}

`;
